src/boot/efifirmware.c

```diff
@@ -0,0 +1,89 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+#include <endian.h>
+
+static bool efifw_validate_header(
+                const void *blob,
+                size_t uki_efifw_len,
+                void *ret_fwid_off,
+                void *ret_payload_off) {
+
+        if ((uintptr_t) blob % alignof(EfiFwHeader) != 0)
+                return false;
+
+        const EfiFwHeader *fw_header;
+
+        /* at least the base size of the header must be in memory */
+        if (uki_efifw_len < sizeof(fw_header->base))
+                return false;
+
+        fw_header = ASSERT_PTR(blob);
+
+        if (fw_header->base.magic != FWHEADERMAGIC)
+                return false;
+
+        uint32_t payload_len = fw_header->base.payload_len;
+        uint32_t header_len  = fw_header->base.header_len;
+        uint32_t fwid_len    = fw_header->base.fwid_len;
+
+        /* header_len must not be malformed */
+        if (header_len < sizeof(fw_header->base))
+                return false;
+
+        size_t fw_base_size;
+
+        /* check for unusually large values of payload_len, header_len or fwid_len */
+        if (!ADD_SAFE(&fw_base_size, header_len, fwid_len) ||
+            !ADD_SAFE(&fw_base_size, fw_base_size, payload_len) )
+                return false;
+
+        if (PTR_TO_SIZE(blob) > SIZE_MAX - fw_base_size)
+                return false;
```

Comment:
this overflow check is not that interesting i#d drop it. instead please validate that `uki_efifw_len >= fw_base_size`. That's enough to validate. because it's ok to assume that the parrt of memory passed into this function is fully valid, and hence cannot possibly overflow.